macros/light-sdk-macros/src/hasher.rs

```diff
@@ -17,28 +17,84 @@ pub(crate) fn hasher(input: ItemStruct) -> Result<TokenStream> {
         }
     };
 
+    fn is_option_type(ty: &syn::Type) -> bool {
+        if let syn::Type::Path(type_path) = ty {
+            if let Some(segment) = type_path.path.segments.last() {
+                return segment.ident == "Option";
+            }
+        }
+        false
+    }
+
     let field_into_bytes_calls = fields
         .named
         .iter()
         .filter(|field| {
-            !field.attrs.iter().any(|attr| {
-                if let Some(attr_ident) = attr.path().get_ident() {
-                    attr_ident == "skip"
-                } else {
-                    false
-                }
-            })
+            !field.attrs.iter().any(|attr| attr.path().is_ident("skip"))
         })
         .map(|field| {
             let field_name = &field.ident;
-            let truncate = field.attrs.iter().any(|attr| {
-                if let Some(attr_ident) = attr.path().get_ident() {
-                    attr_ident == "truncate"
+            let truncate = field.attrs.iter().any(|attr| attr.path().is_ident("truncate"));
+            let nested = field.attrs.iter().any(|attr| attr.path().is_ident("nested"));
+            let is_option = is_option_type(&field.ty);
+
+
+            Ok(if nested {
+                if is_option {
+                    quote! {
+                        match &self.#field_name {
+                            Some(value) => {
+                                let mut nested_hash = ::light_hasher::DataHasher::hash::<::light_hasher::Poseidon>(value)
+                                    .expect("Failed to hash nested field");
+                                nested_hash[0] = 1; // Mark as Some
```

Comment:
That's neat!

What I was thinking about before prepending `0` or `1` to the input of `hash` or `hash_to_bn254_field_size_be`, but actually this trick lets us to do that without additional allocations. I like that.

Reply:
spoke with @ananas-block, we figured the most robust solution is: for nested or truncated options: Some -> hash, None -> 0

Reply:
That sounds great, in that case there is no chance of collision.